### PR TITLE
Add some more documentation.

### PR DESCRIPTION
**Pull request by Hixie**
This changes documentation on code written by @gspencergoog @amirh @xster @HansMuller

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/picker.dart [hunk 1]
```diff
@@ -29,6 +29,10 @@ const double _kForegroundScreenOpacityFraction = 0.7;
 ///    the iOS design specific chrome.
 ///  * <https://developer.apple.com/ios/human-interface-guidelines/controls/pickers/>
 class CupertinoPicker extends StatefulWidget {
+  /// Creates a control used for selecting values.
+  ///
+  /// The [diameterRatio], [backgroundColor], and [itemExtent] arguments must
+  /// not be null. The [itemExtent] must bo greater than zero.
```
Comment (xster) [inline packages/flutter/lib/src/cupertino/picker.dart]: be

File: packages/flutter/lib/src/cupertino/picker.dart [hunk 2]
```diff
@@ -39,6 +43,7 @@ class CupertinoPicker extends StatefulWidget {
     @required this.children,
   }) : assert(diameterRatio != null),
        assert(diameterRatio > 0.0, RenderListWheelViewport.diameterRatioZeroMessage),
+       assert(backgroundColor != null),
```
Comment (xster) [inline packages/flutter/lib/src/cupertino/picker.dart]: This can be null

**Review by xster — COMMENTED** [review]
LG for picker

**Comment by amirh** [comment]
scaffold.dart, framework.dart LGTM

File: packages/flutter/lib/src/cupertino/picker.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter/lib/src/cupertino/picker.dart]: PTAL, i made that case more efficient

**Comment by HansMuller** [comment]
framework.dart changes LGTM

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 3]
```diff
@@ -2681,26 +2681,30 @@ String _concatStrings({
...
-  /// when merged, instead of extending them.
+  /// If [discardParentOrder] is set to true, then the [SemanticsSortOrder.keys]
+  /// will _replace_ the list of keys from the parents when merged, instead of
+  /// extending them.
   SemanticsSortOrder({
     SemanticsSortKey key,
     List<SemanticsSortKey> keys,
     this.discardParentOrder = false,
-  })
-    : assert(key != null || keys != null, 'One of key or keys must be specified.'),
```
Comment (gspencergoog) [inline packages/flutter/lib/src/semantics/semantics.dart]: Damn dartfmt.  I really wish they would fix that one thing: pretty much everything else is working.

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 4]
```diff
@@ -2681,26 +2681,30 @@ String _concatStrings({
...
+       assert(key == null || keys == null, 'Only one of key or keys may be specified.'),
+       keys = key == null ? keys : <SemanticsSortKey>[key];
 
   /// Whether or not this order is to replace the keys above it in the
   /// semantics tree, or to be appended to them.
   final bool discardParentOrder;
 
+  /// The keys that should be used to sort this node.
+  ///
+  /// Typically only one key is provided, using the constructor's `key` argument.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/semantics/semantics.dart]: Well, I wouldn't say it's typical. There are two main ways to use it: either specify the entire string of keys in one go, replacing whatever the hierarchy built, or specify one at each level of the widget hierarchy.

I'd say something like "If only one key is needed, using the constructor's `key` argument is less verbose."

**Review by gspencergoog — APPROVED** [review]

**Comment by gspencergoog** [comment]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 3 above)
Comment (Hixie) [reply packages/flutter/lib/src/semantics/semantics.dart]: file a bug on dartfmt, see what bob says :-)

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 4 above)
Comment (Hixie) [reply packages/flutter/lib/src/semantics/semantics.dart]: I guess I don't really understand how this works. Can you suggest some sample code I could include here to explain what "keys" does?

**Comment by xster** [comment]
LGTM

**Comment by Hixie** [comment]
Thanks for the reviews.

**Pull request merged**